internal/core/service/user_service.go

```diff
@@ -0,0 +1,158 @@
+package service
+
+import (
+	"errors"
+	"github.com/gin-gonic/gin"
+	"github.com/viniciusgferreira/ps-tag-onboarding-go/internal/core/domain/models"
+	"github.com/viniciusgferreira/ps-tag-onboarding-go/internal/core/ports"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"regexp"
+)
+
+type UserService struct {
+	repo ports.UserRepository
+}
+
+type InvalidUserResponse struct {
+	Name    string   `json:"error"`
+	Details []string `json:"details,omitempty"`
+}
+
+func (r InvalidUserResponse) Error() string {
+	return r.Name
+}
+
+func (r InvalidUserResponse) New(validationErrors []error) InvalidUserResponse {
+	details := make([]string, len(validationErrors))
+	for i, err := range validationErrors {
+		details[i] = err.Error()
+	}
+	return InvalidUserResponse{
+		Name:    "User did not pass validation",
+		Details: details,
+	}
+}
+
+var (
+	ErrUserNotFound      = errors.New("User not found")
+	ErrUserAlreadyExists = errors.New("User with same first and last name already exists")
+	ErrInvalidAge        = errors.New("User must be at least 18 years old")
+	ErrInvalidName       = errors.New("User first and last name cannot be empty")
+	ErrInvalidEmail      = errors.New("Invalid email format")
+	ErrInvalidID         = errors.New("Invalid id (ObjectID)")
+)
+
+func New(repo ports.UserRepository) *UserService {
+	return &UserService{repo: repo}
+}
+func (s *UserService) Find(ctx *gin.Context, id string) (*models.User, error) {
+	err := s.validateID(id)
+	if err != nil {
+		return nil, err
+	}
+	user, err := s.repo.FindById(ctx, id)
+	if err != nil {
+		if errors.Is(err, mongo.ErrNoDocuments) {
+			return nil, ErrUserNotFound
+		}
+		return nil, err
+	}
+	return user, nil
+}
+
+func (s *UserService) Save(ctx *gin.Context, u models.User) (*models.User, error) {
+	validationErrors := s.Validate(u)
+	if validationErrors != nil {
+		errorsReponse := InvalidUserResponse{}
+		return nil, errorsReponse.New(validationErrors)
+	}
+	exists, err := s.repo.ExistsByFirstNameAndLastName(ctx, u.FirstName, u.LastName)
+	if err != nil {
+		return nil, err
+	}
+	if exists {
+		return nil, ErrUserAlreadyExists
```

Comment:
Please, double check in the java project the status code returned for a duplicate user and if it should or not  append in the validationErrors